### ISSUE TITLE
The `stop_gradient` op in FastGradientMethod is wrong

### ISSUE THREAD

**Issue by goodfeli** [tensorflow/cleverhans#243]
The `stop_gradient` op in FastGradientMethod should only be present for the max norm and in that case it should wrap only the sign op, not the addition of the perturbation to the input. The `stop_gradient` op should reduce computational cost but should not change the numeric value of any derivatives.

**Comment by AlexeyKurakin** [comment]
Just to clarify, why  `adv = tf.stop_gradient(x + eps * tf.sign(g))` is wrong?
I always had an impression that we want to treat adversarial examples as a constant (i.e. don't backprop through it for adversarial training) and that's the purpose of tf.stop_gradient

**Comment by goodfeli** [comment]
It's fine to define adversarial training to treat the adversarial examples as constant, but that should be done in the training code, rather than the attack definition code. You could also define adversarial training as backpropagating through the adversarial example construction process. (We tried that for rotational adversarial examples in "Explaining and Harnessing Adversarial Examples"). In theory, that lets the defender anticipate the way the attacker will change in the future. For FGSM, it has no effect, because the gradient of the attacker is zero anyway.

**Comment by goodfeli** [comment]
There are also two issues here: differentiability of the perturbation, and differentiability of the input `x`. For adversarial training, `x` comes from the dataset, so it doesn't really matter whether it is wrapped in a `stop_gradient` or not. In other contexts, like if `x` were a sample from a GAN and we want to make an adversarial perturbation that fools the discriminator, we'd want to be able to backpropagate through `x` to train the GAN.
